contracts/utils/cryptography/signers/MultiSignerERC7913Weighted.sol

```diff
@@ -0,0 +1,178 @@
+     * default weight of 1. See {_signerWeight}.
+     */
+    event ERC7913SignerWeightChanged(bytes indexed signer, uint64 weight);
+
+    /// @dev Thrown when a signer's weight is invalid.
+    error MultiSignerERC7913WeightedInvalidWeight(bytes signer, uint64 weight);
+
+    /// @dev Thrown when the threshold is unreachable.
+    error MultiSignerERC7913WeightedMismatchedLength();
+
+    /// @dev Gets the weight of a signer. Returns 0 if the signer is not authorized.
+    function signerWeight(bytes memory signer) public view virtual returns (uint64) {
+        return uint64(Math.ternary(isSigner(signer), _signerWeight(signer), 0)); // Safe cast, both uint64
+    }
+
+    /// @dev Gets the total weight of all signers.
+    function totalWeight() public view virtual returns (uint64) {
+        return _weightSigners(getSigners(0, type(uint64).max));
+    }
+
+    /**
+     * @dev Gets the weight of the current signer. Returns 1 if not explicitly set.
+     *
+     * NOTE: This internal function doesn't check if the signer is authorized.
+     */
+    function _signerWeight(bytes memory signer) internal view virtual returns (uint64) {
+        return uint64(Math.max(_weights[signer], 1)); // Safe cast, both uint64
+    }
+
+    /**
+     * @dev Sets weights for multiple signers at once. Internal version without access control.
+     *
+     * Requirements:
+     *
+     * - `signers` and `weights` arrays must have the same length. Reverts with {MultiSignerERC7913WeightedMismatchedLength} on mismatch.
+     * - Each signer must exist in the set of authorized signers. Reverts with {MultiSignerERC7913NonexistentSigner} if not.
+     * - Each weight must be greater than 0. Reverts with {MultiSignerERC7913WeightedInvalidWeight} if not.
+     * - See {_validateReachableThreshold} for the threshold validation.
+     *
+     * Emits {ERC7913SignerWeightChanged} for each signer.
+     */
+    function _setSignerWeights(bytes[] memory signers, uint64[] memory newWeights) internal virtual {
+        require(signers.length == newWeights.length, MultiSignerERC7913WeightedMismatchedLength());
+
+        for (uint256 i = 0; i < signers.length; ++i) {
+            bytes memory signer = signers[i];
+            uint64 newWeight = newWeights[i];
+            require(isSigner(signer), MultiSignerERC7913NonexistentSigner(signer));
+            require(newWeight > 0, MultiSignerERC7913WeightedInvalidWeight(signer, newWeight));
+        }
+
+        _unsafeSetSignerWeights(signers, newWeights);
+        _validateReachableThreshold();
+    }
+
+    /**
+     * @dev See {MultiSignerERC7913-_removeSigners}.
+     *
+     * Emits {ERC7913SignerWeightChanged} for each removed signer.
+     */
+    function _removeSigners(bytes[] memory oldSigners) internal virtual override {
+        // Clean up weights for removed signers
+        _unsafeSetSignerWeights(oldSigners, new uint64[](oldSigners.length));
```

Comment:
this allocated memory that is not really used. Given that `_unsafeSetSignerWeights` is private anyway, maybe it would be better to do:
```suggestion
        for (uint256 i = 0; i < oldSigners.length; ++i) {
            delete _weights[oldSigners[i]];
            emit ERC7913SignerWeightChanged(oldSigners[i], 0);
        }
```